File: compiler/C/lexer.c

#include <stdio.h>
#include <stdlib.h>

#include "debug.h"
#include "lexer.h"
#include "ArrayList.h"

// read a file and return its contents
char* readFile(char* filename){
    FILE* theFile = fopen(filename, "r");

    if (theFile == NULL){
        panic("ERROR: Could not open the input file.\n");
    }

    dbg("Input file opened successfully!\n");

    // go to end of file and get position to get size
    fseek(theFile, 0, SEEK_END);
    long fileLength = ftell(theFile);

    // allocate buffer to store file contents
    char* fileBuffer = (char*) malloc(sizeof(char) * (fileLength + 1));

    // go back to the start of the file and read it into the buffer
    fseek(theFile, 0, SEEK_SET);
    size_t bytesWritten = fread(fileBuffer, sizeof(size_t), fileLength/sizeof(size_t), theFile) * sizeof(size_t);

    int remainingBytes = fileLength % sizeof(size_t);
    if (remainingBytes != 0){
        fread(fileBuffer + bytesWritten, sizeof(char), remainingBytes, theFile);
    }
    fclose(theFile);

    fileBuffer[(size_t) fileLength] = '\0';

    dbg("File has successfully been read into memory...\n");
    dbg("File contents:\n");
    dbg("==============\n");
    dbg(fileBuffer);
    dbg("==============\n");

    return fileBuffer;
}














///////////////////////////////////////////////////
// public functions

char** performLexicalAnalysis(char* filename){

    // get the file contents
    dbg("Reading the input file...\n");
    char* fileContents = readFile(filename);

    // break the file up into tokens
    

    return NULL;
}

File: compiler/python/lexer.py

from debug import *
from token import *

class Lexer:

    def __init__(self, filename:str):
        self.filename = filename

        self.file_data = self.open_file(self.filename)

        self.tokens = self.tokenize(self.file_data)
        for token in self.tokens:
            token.filename = self.filename

        # combine postfix and prefix operators while there are still spaces
        i = 0
        n = len(self.tokens)
        while i < n:
            if self.tokens[i].token in ["+", "-"]:
                if i + 1 < n and self.tokens[i].token == self.tokens[i+1].token:
                    self.tokens[i].token += self.tokens[i+1].token
                    del self.tokens[i+1]
                    n -= 1
            i += 1


    def open_file(self, filename:str) -> str:
        dbg(f"Opening file {filename}...")
        try:
            f = open(filename, 'r')
        except:
            panic(f"Unable to open file ({filename})")

        dbg(f"File read successfully!")
        return f.read()


    def tokenize(self, file_data:str) -> list[Token]:
        i = 0
        n = len(file_data)

        breakChars = {"~", "!", "#", "%", "^", "&", "*", "(", ")", "-", "+", "=", "{", "}", "[", "]", "|", '\\', "'", '"', ';', ":", "/", "?", ".", ",", "<", ">", '\n', '\t', ' '}

        result = []
        line_number = 1

        current_token = ""
        while i < n:

            if file_data[i] in breakChars:
                if file_data[i] == "\n":
                    line_number += 1

                if len(current_token) > 0:
                    result.append(Token(current_token, line_number, ""))

                result.append(Token(file_data[i], line_number, ""))

                current_token = ""

            else:
                current_token += file_data[i]

            i += 1

        return result





